include/cetl/variable_length_array.hpp

```diff
@@ -0,0 +1,2065 @@
+    /// @param alloc Allocator to use for all allocations.
+    /// @param max_size_max Clamping value for the maximum size of this array. That is,
+    /// cetl::VariableLengthArray::max_size() will return
+    /// `std::min(max_size_max, std::allocator_traits<allocator_type>::max_size(alloc))`
+    explicit VariableLengthArray(const allocator_type& alloc, size_type max_size_max) noexcept
+        : Base(alloc, nullptr, 0, 0, max_size_max)
+    {
+    }
+
+    /// Initializer syntax constructor.
+    /// @param l Initializer list of elements to copy into the array.
+    /// @param alloc Allocator to use for all allocations.
+    /// @param max_size_max Clamping value for the maximum size of this array. That is,
+    /// cetl::VariableLengthArray::max_size() will return
+    /// `std::min(max_size_max, std::allocator_traits<allocator_type>::max_size(alloc))`
+    VariableLengthArray(std::initializer_list<value_type> l,
+                        const allocator_type&             alloc,
+                        size_type                         max_size_max = std::numeric_limits<size_type>::max())
+        : Base(alloc, nullptr, 0, 0, max_size_max)
+    {
+        reserve(l.size());
+        size_ = Base::fast_copy_construct(data_, capacity_, l.begin(), l.size(), alloc_);
+    }
+
+    /// Range constructor.
+    /// @tparam InputIt The type of the range's iterators.
+    /// @param first    The beginning of the range.
+    /// @param last     The end of the range.
+    /// @param alloc    Allocator to use for all allocations.
+    /// @param max_size_max Clamping value for the maximum size of this array. That is,
+    /// cetl::VariableLengthArray::max_size() will return `std::min(max_size_max,
+    /// std::allocator_traits<allocator_type>::max_size(alloc))`
+    template <class InputIt>
+    VariableLengthArray(InputIt               first,
+                        InputIt               last,
+                        const allocator_type& alloc,
+                        size_type             max_size_max = std::numeric_limits<size_type>::max())
+        : Base(alloc, nullptr, 0, 0, max_size_max)
+    {
+        if (last >= first)
+        {
+            reserve(static_cast<std::size_t>(last - first));
+            if (nullptr != data_)
+            {
+                size_ =
+                    Base::fast_copy_construct(data_, capacity_, first, static_cast<std::size_t>(last - first), alloc_);
+            }
+        }
+    }
+
+    //
+    // Rule of Five.
+    //
+    VariableLengthArray(const VariableLengthArray& rhs)
+        : Base(rhs)
+    {
+        Base::reserve(rhs.size(), rhs.max_size());
+        size_ = Base::fast_copy_construct(data_, capacity_, rhs.data_, rhs.size_, alloc_);
+    }
+
+    VariableLengthArray& operator=(const VariableLengthArray& rhs)
+    {
+        Base::copy_assign_from(rhs, rhs.max_size());
+        return *this;
+    }
+
+    VariableLengthArray(VariableLengthArray&& rhs) noexcept
```

Comment:
std::vector has copy/move constructors with a second parameter that allows passing in a different allocator.   Don't we want that for VLA too?

Reply:
If we need it I can add.

Reply:
I'm planning on using it.  If you want to add in a later pass, that's fine too.

Reply:
Added the copy version since that was trivial. The move version is more work so I added #31